crates/revm/src/instructions/arithmetic.rs

```diff
@@ -18,60 +19,45 @@ pub fn sdiv(op1: U256, op2: U256) -> U256 {
 }
 
 pub fn rem(op1: U256, op2: U256) -> U256 {
-    if op2.is_zero() {
-        U256::zero()
+    if op2 == U256::ZERO {
+        U256::ZERO
     } else {
         op1.rem(op2)
     }
 }
 
 pub fn smod(op1: U256, op2: U256) -> U256 {
-    if op2.is_zero() {
-        U256::zero()
+    if op2 == U256::ZERO {
+        U256::ZERO
     } else {
         i256_mod(op1, op2)
     }
 }
 
 pub fn addmod(op1: U256, op2: U256, op3: U256) -> U256 {
-    if op3.is_zero() {
-        U256::zero()
-    } else {
-        let op1: U512 = op1.into();
-        let op2: U512 = op2.into();
-        let op3: U512 = op3.into();
-        let v = (op1 + op2) % op3;
-        v.try_into()
-            .expect("op3 is less than U256::MAX, thus it never overflows; qed")
-    }
+    op1.add_mod(op2, op3)
 }
 
 pub fn mulmod(op1: U256, op2: U256, op3: U256) -> U256 {
-    if op3.is_zero() {
-        U256::zero()
-    } else {
-        let op1: U512 = op1.into();
-        let op2: U512 = op2.into();
-        let op3: U512 = op3.into();
-        let v = (op1 * op2) % op3;
-        v.try_into()
-            .expect("op3 is less than U256::MAX, thus it never overflows; qed")
-    }
+    op1.mul_mod(op2, op3)
 }
 
 pub fn exp(op1: U256, op2: U256) -> U256 {
+    let mut result = U256::from(1);
     let mut op1 = op1;
     let mut op2 = op2;
-    let mut r: U256 = 1.into();
-
-    while op2 != 0.into() {
-        if op2 & 1.into() != 0.into() {
-            r = r.overflowing_mul(op1).0;
+    // Exponentiation by squaring
+    while op2 > U256::ZERO {
+        // Multiply by base
+        if op2.as_limbs()[0] & 1 == 1 {
+            result = result.mul(op1);
         }
+
+        // Square base
+        op1 = op1.mul(op1);
         op2 >>= 1;
-        op1 = op1.overflowing_mul(op1).0;
     }
-    r
+    result
```

Comment:
There is `wrapping_pow` but it takes a `usize` argument. Maybe I should add a function taking a Uint.